JASP-Engine/JASP/R/exploratoryfactoranalysis.R

```diff
@@ -27,379 +28,315 @@
+      keep <- c(keep, results[["screePlot"]][["data"]])
+    }
+  }
   ## TEMP DEBUG THING:
   # save(dataset,results,init,options,perform,callback,...,file = "/Users/sachaepskamp/Dropbox/work/JASP/Rcodes/JASPinit.RData")
+
   
-  
-  #### META
-  meta <- list(
-    list(name = "title", type = "title")
+  # # Dummies:
+  # status <- list(ready=TRUE, error=error)
+
+  #save state
+  state <- list(
+    options = options,
+    analysisResults = analysisResults,
+    pathDiagram = results[["pathDiagram"]],
+    screePlot = results[["screePlot"]]
   )
-  
-  if (isTRUE(options$incl_loadings)){
-    meta[[length(meta)+1]] <- list(name = "factorLoadings", type = "table")
-    
-  }
-  
-  if (isTRUE(options$incl_correlations)){
-    meta[[length(meta)+1]] <- list(name = "factorCorrelations", type = "table")
-  }
-  
-  if (isTRUE(options$incl_GoF)){
-    meta[[length(meta)+1]] <- list(name = "goodnessOfFit", type = "table")
-  }
-  if (isTRUE(options$incl_fitIndices)){
-    meta[[length(meta)+1]] <- list(name = "fitMeasures", type = "table")
-  }
-  if (isTRUE(options$incl_pathDiagram)){
-    meta[[length(meta)+1]] <- list(name = "pathDiagram", type = "image")
-  }
-  if (isTRUE(options$incl_screePlot)){
-    meta[[length(meta)+1]] <- list(name = "screePlot", type = "image")
-  }
-  
-  
-  
-  results[[".meta"]] <- meta
-  
-  # Dummies:
-  status <- list(ready=TRUE, error=error,errorMessage=errorMessage)
-  
-  if (status$error == TRUE){
-    results[["factorLoadings"]][["error"]] <-  list(errorType="badData", errorMessage=status$errorMessage)
-    results[["factorCorrelations"]][["error"]] <-  list(errorType="badData")
-    results[["goodnessOfFit"]][["error"]] <-  list(errorType="badData")
-    results[["fitMeasures"]][["error"]] <-  list(errorType="badData")
-    results[["pathDiagram"]][["error"]] <-  list(errorType="badData")
-    results[["screePlot"]][["error"]] <-  list(errorType="badData")
+
+  attr(state, "key") <- stateKey
+
+  if (perform == "run") {
+
+    return(list(results = results, status = "complete", state = state, keep = keep))
+
   } else {
-    results[["factorLoadings"]][["error"]] <- NULL
-    results[["factorCorrelations"]][["error"]] <-  NULL
-    results[["goodnessOfFit"]][["error"]] <-  NULL
-    results[["fitMeasures"]][["error"]] <-  NULL
-    results[["pathDiagram"]][["error"]] <-  NULL
-    results[["screePlot"]][["error"]] <-  NULL
+
+    return(list(results = results, status = "inited", state = state, keep = keep))
   }
-  
-  if (perform == "run" && status$ready) {
-    state <- list(options=options,analysisResults=analysisResults,nFactor=nFactor,results=results,complete=TRUE)
-    retList <- list(results=results, status="complete", state=state)
+}
 
-  } else {
-    state <- list(options=options,nFactor=nFactor,analysisResults=analysisResults,results=results,complete=FALSE)
-    
-    retList <- list(results=results, status="inited",state=state)
+.estimateNFactorEFA <- function(dataset, options, perform){
+
```

Comment:
@hummyhummy is there also a function like this for pca?

Reply:
Yes, .estimateNFactorPCA